src/bootchooser.c

```diff
@@ -211,6 +211,73 @@ static gboolean barebox_set_state(RaucSlot *slot, gboolean good, GError **error)
 	return res;
 }
 
+/* Get slot marked as primary one */
+static RaucSlot* barebox_get_primary(GError **error) {
+	RaucSlot *slot;
+	GHashTableIter iter;
+	RaucSlot *primary = NULL;
+	guint32 top_prio = 0;
+	GError *ierror = NULL;
+	gboolean res;
+
+	g_hash_table_iter_init(&iter, r_context()->config->slots);
+
+	while (g_hash_table_iter_next(&iter, NULL, (gpointer *)&slot)) {
+		BareboxSlotState state;
+
+		if (!slot->bootname)
+			continue;
+
+		res = barebox_state_get(slot->bootname, &state, &ierror);
+		if (!res) {
+			g_debug("%s", ierror->message);
+			g_clear_error(&ierror);
+			continue;
+		}
+
+		if (state.attempts == 0)
+			continue;
+
+		/* We search for the slot with highest priority */
+		if (state.prio > top_prio) {
+			primary = slot;
+			top_prio = state.prio;
+		}
+	}
+
+	if (!primary) {
+		g_set_error_literal(
+				error,
+				R_BOOTCHOOSER_ERROR,
+				R_BOOTCHOOSER_ERROR_PARSE_FAILED,
+				"Unable to obtain primary element");
+	}
+
+	return primary;
+}
+
+/* We assume a slot to be 'good' if its priority is > 0 AND its remaining
+ * attempts counter is > 0 */
```

Comment:
This may not be correct for all configurations of the barebox bootchooser, but we probably can't detect that from the running system.